SyntaxKit/RegularExpression.swift

```diff
@@ -0,0 +1,139 @@
+        static let anchored = MatchingOptions(rawValue: NSRegularExpression.MatchingOptions.anchored.rawValue)
+        static let withTransparentBounds = MatchingOptions(rawValue: NSRegularExpression.MatchingOptions.withTransparentBounds.rawValue)
+        static let withoutAnchoringBounds = MatchingOptions(rawValue: NSRegularExpression.MatchingOptions.withoutAnchoringBounds.rawValue)
+    }
+
+    struct MatchingFlags : OptionSet {
+        let rawValue: UInt
+        init(rawValue: UInt) {
+            self.rawValue = rawValue
+        }
+        static var progress = MatchingFlags(rawValue: NSRegularExpression.MatchingFlags.progress.rawValue)
+        static var completed = MatchingFlags(rawValue: NSRegularExpression.MatchingFlags.completed.rawValue)
+        static var hitEnd = MatchingFlags(rawValue: NSRegularExpression.MatchingFlags.hitEnd.rawValue)
+        static var requiredEnd = MatchingFlags(rawValue: NSRegularExpression.MatchingFlags.requiredEnd.rawValue)
+        static var internalError = MatchingFlags(rawValue: NSRegularExpression.MatchingFlags.internalError.rawValue)
+    }
+}
+
+internal extension RegularExpression {
+
+    var numberOfCaptureGroups: Int {
+        return _expression?.numberOfCaptureGroups ?? 0
+    }
+
+    static func escapedPattern(for string: String) -> String {
+        return NSRegularExpression.escapedPattern(for: string)
+    }
+
+}
+
+internal extension RegularExpression {
+
+    func enumerateMatches(in string: String, options: RegularExpression.MatchingOptions = [], range: NSRange, using block: (NSTextCheckingResult?, RegularExpression.MatchingFlags, UnsafeMutablePointer<ObjCBool>) -> Void) {
+        _expression?.enumerateMatches(in: string, options: NSRegularExpression.MatchingOptions(rawValue: options.rawValue), range: range) { (result, flags, stop) in
+            block(result, RegularExpression.MatchingFlags(rawValue: flags.rawValue), stop)
+        }
+    }
+
+    func matches(in string: String, options: RegularExpression.MatchingOptions = [], range: NSRange) -> [NSTextCheckingResult] {
+        return _expression?.matches(in: string, options: NSRegularExpression.MatchingOptions(rawValue: options.rawValue), range: range) ?? []
+    }
+
+    func numberOfMatches(in string: String, options: RegularExpression.MatchingOptions = [], range: NSRange) -> Int {
+        return _expression?.numberOfMatches(in: string, options: NSRegularExpression.MatchingOptions(rawValue: options.rawValue), range: range) ?? 0
+    }
+
+    func firstMatch(in string: String, options: RegularExpression.MatchingOptions = [], range: NSRange) -> NSTextCheckingResult? {
+        return _expression?.firstMatch(in: string, options: NSRegularExpression.MatchingOptions(rawValue: options.rawValue), range: range)
+    }
+
+    func rangeOfFirstMatch(in string: String, options: RegularExpression.MatchingOptions = [], range: NSRange) -> NSRange {
```

Comment:
This would be clearer if written as

```
// guard expression
return _expression.rangeOfFirstMatch(in: string, options: NSRegularExpression.MatchingOptions(rawValue: options.rawValue), range: range)
```

 (since the rangeOfFirstMatch will never return nil). That way we know that the interface is not changed.

Reply:
In general we want to keep the exact same interface as NSRegularExpression.